### PR TITLE
feat: add support for bsc parlia consensus

### PR DESCRIPTION
**Pull request by pythonberg1997**
TODO

### CODE REVIEW

File: crates/interfaces/Cargo.toml [hunk 1]
```diff
@@ -40,6 +40,7 @@ tokio = { workspace = true, features = ["full"] }
 secp256k1 = { workspace = true, features = ["alloc", "recovery", "rand"] }
 
 [features]
+bsc = []
```
Comment (unclezoro) [inline crates/interfaces/Cargo.toml]: TODO?

File: crates/interfaces/src/consensus.rs [hunk 2]
```diff
@@ -246,9 +246,123 @@ pub enum ConsensusError {
     /// Error type transparently wrapping HeaderValidationError.
     #[error(transparent)]
     HeaderValidationError(#[from] HeaderValidationError),
+
+    /// Error type transparently wrapping ParliaConsensusError.
+    #[error(transparent)]
+    ParliaConsensusError(#[from] ParliaConsensusError),
```
Comment (unclezoro) [inline crates/interfaces/src/consensus.rs]: I think it should be more concrete error, but not `Parliaxxx` error, for example `InvalidDifficulty` error is kind of `HeaderValidationError`, let us try to reuse current trait and add more Enum type to current errors.

File: crates/revm/src/factory.rs [hunk 3]
```diff
@@ -5,23 +5,26 @@ use crate::{
...
 /// Factory for creating [EVMProcessor].
 #[derive(Clone, Debug)]
-pub struct EvmProcessorFactory<EvmConfig> {
+pub struct EvmProcessorFactory<EvmConfig, P> {
     chain_spec: Arc<ChainSpec>,
     stack: Option<InspectorStack>,
     /// Type that defines how the produced EVM should be configured.
     evm_config: EvmConfig,
+    /// Parlia consensus instance
+    parlia_consensus: Option<Arc<Parlia<P>>>,
```
Comment (unclezoro) [inline crates/revm/src/factory.rs]: This seems not a good pattern, I did not see anywhere use this?

File: crates/revm/src/processor.rs [hunk 4]
```diff
@@ -60,9 +72,11 @@ pub struct EVMProcessor<'a, EvmConfig> {
     pub(crate) stats: BlockExecutorStats,
     /// The type that is able to configure the EVM environment.
     _evm_config: EvmConfig,
+
+    parlia_consensus: Option<Arc<Parlia<P>>>,
```
Comment (unclezoro) [inline crates/revm/src/processor.rs]: It seems not a good pattern. 

If it is used as a flag, we can replace with feature?

File: crates/revm/src/processor.rs [hunk 5]
```diff
@@ -287,11 +308,385 @@ where
...
+
+        if self.batch_record.first_block().is_none() {
+            self.batch_record.set_first_block(block.number);
+        }
+
+        Ok(receipts)
+    }
+
+    #[cfg(feature = "bsc")]
+    pub fn finalize(
```
Comment (unclezoro) [inline crates/revm/src/processor.rs]: I think it is better to move to consensus.

File: crates/interfaces/Cargo.toml (hunk 1 above)
Comment (pythonberg1997) [reply crates/interfaces/Cargo.toml]: No. `[]` is legal. This means this feature requires no additional crate.

File: crates/interfaces/src/consensus.rs (hunk 2 above)
Comment (pythonberg1997) [reply crates/interfaces/src/consensus.rs]: Sure. This is a transparent error. I do implement a `ParliaConsensusError` enum struct.

File: crates/revm/src/processor.rs (hunk 4 above)
Comment (pythonberg1997) [reply crates/revm/src/processor.rs]: Done

File: crates/consensus/auto-seal/src/lib.rs [hunk 6]
```diff
@@ -337,10 +337,10 @@ impl StorageInner {
     /// Executes the block with the given block and senders, on the provided [EVMProcessor].
     ///
     /// This returns the poststate from execution and post-block changes, as well as the gas used.
-    pub(crate) fn execute<EvmConfig>(
+    pub(crate) fn execute<EvmConfig, P>(
         &mut self,
         block: &BlockWithSenders,
-        executor: &mut EVMProcessor<'_, EvmConfig>,
+        executor: &mut EVMProcessor<'_, EvmConfig, P>,
```
Comment (unclezoro) [inline crates/consensus/auto-seal/src/lib.rs]: I think `EVMProcessor` is a very high level trait,  prefer not to inject the `Parlia` engine into it.

File: crates/consensus/parlia/src/go_rng.rs [hunk 7]
```diff
@@ -0,0 +1,857 @@
+/// Uniform distribution
+/// algorithm by
+/// DP Mitchell and JA Reeds
+
+const RNG_LEN: usize = 607;
```
Comment (unclezoro) [inline crates/consensus/parlia/src/go_rng.rs]: move this file to a more common packages?

File: crates/consensus/parlia/src/go_rng.rs [hunk 8]
```diff
@@ -0,0 +1,857 @@
+/// Uniform distribution
+/// algorithm by
+/// DP Mitchell and JA Reeds
```
Comment (unclezoro) [inline crates/consensus/parlia/src/go_rng.rs]: can you comment the detailed go implementation lib?

Comment (unclezoro) [reply crates/consensus/parlia/src/go_rng.rs]: How does akula handle this random library?

File: crates/consensus/parlia/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,1215 @@
...
+}
+
+#[derive(Clone, Debug)]
+pub struct ParliaConfig {
+    epoch: u64,
+    period: u64,
+}
+
+/// BSC parlia consensus implementation
+pub struct Parlia<P> {
```
Comment (unclezoro) [inline crates/consensus/parlia/src/lib.rs]: Should't `P` have some trait?

File: crates/consensus/parlia/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,1215 @@
...
+        if raw.is_empty() {
+            return Ok(None);
+        }
+
+        Ok(Some(
+            Decodable::decode(&mut raw).map_err(|_| ParliaConsensusError::ABIDecodeInnerError)?,
+        ))
+    }
+
+    pub fn get_snapshot_from_cache(&self, hash: &B256) -> Option<Snapshot> {
```
Comment (unclezoro) [inline crates/consensus/parlia/src/lib.rs]: not used yet?

File: crates/consensus/parlia/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,1215 @@
...
+            BACKOFF_TIME_OF_INITIAL + (backoff_steps[index] as u64) * BACKOFF_TIME_OF_WIGGLE;
+        // If the current validator has recently signed, reduce initial delay.
+        if recents.get(&snap.inturn_validator()).is_some() {
+            delay -= BACKOFF_TIME_OF_INITIAL;
+        }
+        delay
+    }
+}
+
+impl<P> Parlia<P> {
```
Comment (unclezoro) [inline crates/consensus/parlia/src/lib.rs]: I can see many `impl<P> Parlia<P>`, can we structure them better in a more readable way.

File: crates/consensus/parlia/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,1215 @@
...
+    }
+
+    pub fn get_max_elected_validators(&self) -> (Address, Bytes) {
+        let function =
+            self.stake_hub_abi.function("maxElectedValidators").unwrap().first().unwrap();
+
+        (*STAKE_HUB_CONTRACT, Bytes::from(function.abi_encode_input(&[]).unwrap()))
+    }
+
+    pub fn unpack_data_into_max_elected_validators(&self, data: &[u8]) -> U256 {
```
Comment (unclezoro) [inline crates/consensus/parlia/src/lib.rs]: are these `unpack` function still usable?

File: crates/revm/src/factory.rs [hunk 13]
```diff
@@ -5,23 +5,37 @@ use crate::{
...
 /// Factory for creating [EVMProcessor].
 #[derive(Clone, Debug)]
-pub struct EvmProcessorFactory<EvmConfig> {
+pub struct EvmProcessorFactory<EvmConfig, P> {
     chain_spec: Arc<ChainSpec>,
     stack: Option<InspectorStack>,
     /// Type that defines how the produced EVM should be configured.
     evm_config: EvmConfig,
+
+    _phantom: PhantomData<P>,
```
Comment (unclezoro) [inline crates/revm/src/factory.rs]: `_phantom` seems useless?

File: crates/revm/src/factory.rs [hunk 14]
```diff
@@ -5,23 +5,37 @@ use crate::{
...
     /// Create new factory
     pub fn new(chain_spec: Arc<ChainSpec>, evm_config: EvmConfig) -> Self {
-        Self { chain_spec, stack: None, evm_config }
+        Self {
+            chain_spec,
+            stack: None,
+            evm_config,
+            _phantom: PhantomData::default(),
+            #[cfg(feature = "bsc")]
+            parlia_cfg: None,
```
Comment (unclezoro) [inline crates/revm/src/factory.rs]: I think a factory should not have `parlia_cfg`, but it can generate a `Processor` that have  `parlia_cfg`?

Comment (unclezoro) [reply crates/revm/src/factory.rs]: prefer to use trait instead of concreat `parlia_cfg`?

File: crates/revm/src/factory.rs [hunk 15]
```diff
@@ -35,25 +49,36 @@ impl<EvmConfig> EvmProcessorFactory<EvmConfig> {
...
-        let mut evm = EVMProcessor::new_with_db(
+        let mut evm = EVMProcessor::<EvmConfig, P>::new_with_db(
             self.chain_spec.clone(),
             database_state,
             self.evm_config.clone(),
         );
         if let Some(stack) = &self.stack {
             evm.set_stack(stack.clone());
         }
+        #[cfg(feature = "bsc")]
```
Comment (unclezoro) [inline crates/revm/src/factory.rs]: not good code pattern, if more chain support, it will be chaos,

File: crates/revm/src/processor.rs [hunk 16]
```diff
@@ -382,6 +794,91 @@ where
...
+            .active_at_timestamp(block.timestamp)
+        {
+            let _parent =
+                self.parlia_consensus.get_header_by_hash(block.number - 1, block.parent_hash)?;
+            // apply system contract upgrade
+            todo!()
+        }
+
+        // perf: do not execute empty blocks
+        if block.body.is_empty() {
```
Comment (unclezoro) [inline crates/revm/src/processor.rs]: even an empty block, may need to consensus finalize (like breath block and so on)

File: crates/revm/src/processor.rs [hunk 17]
```diff
@@ -287,11 +321,388 @@ where
...
     }
+
+    #[cfg(feature = "bsc")]
+    pub(crate) fn execute_inner(
+        &mut self,
+        block: &BlockWithSenders,
+        total_difficulty: U256,
+    ) -> Result<Vec<Receipt>, BlockExecutionError> {
+        self.init_env(&block.header, total_difficulty);
+        let (mut system_tx, mut receipts, mut cumulative_gas_used) =
```
Comment (unclezoro) [inline crates/revm/src/processor.rs]: ==> `system_txs`?

File: crates/revm/src/processor.rs [hunk 18]
```diff
@@ -287,11 +321,388 @@ where
...
+        header: &Header,
+        system_txs: &mut Vec<&TransactionSigned>,
+        receipts: &mut Vec<Receipt>,
+        cumulative_gas_used: &mut u64,
+    ) -> Result<(), BlockExecutionError> {
+        let number = header.number;
+        let validator = header.beneficiary;
+        let parent = self.parlia_consensus.get_header_by_hash(header.number, header.parent_hash)?;
+
+        // The snapshot should be ready after the header stage
```
Comment (unclezoro) [inline crates/revm/src/processor.rs]: can we wrap most of the following code into the conseneus parlia crate?

File: crates/revm/src/processor.rs [hunk 19]
```diff
@@ -287,11 +321,388 @@ where
...
+                        cumulative_gas_used,
+                    )?;
+                }
+
+                block_reward -= reward_to_system;
+            }
+        }
+
+        let nonce = self.db_mut().basic(validator).unwrap().unwrap().nonce;
+        self.parlia_consensus.distribute_to_validator(nonce, validator, block_reward)?;
```
Comment (unclezoro) [inline crates/revm/src/processor.rs]: need todo self.transact_system_tx?

File: crates/stages/src/stages/headers.rs [hunk 20]
```diff
@@ -79,6 +86,24 @@ where
         consensus: Arc<dyn Consensus>,
         etl_config: EtlConfig,
     ) -> Self {
+        #[cfg(feature = "bsc")]
```
Comment (unclezoro) [inline crates/stages/src/stages/headers.rs]: is it a good pattern to use the feature inside a function?

File: crates/revm/src/factory.rs (hunk 3 above)
Comment (pythonberg1997) [reply crates/revm/src/factory.rs]: Fixed

File: crates/consensus/parlia/src/go_rng.rs (hunk 8 above)
Comment (pythonberg1997) [reply crates/consensus/parlia/src/go_rng.rs]: I copied from akula

File: crates/consensus/parlia/src/go_rng.rs (hunk 7 above)
Comment (pythonberg1997) [reply crates/consensus/parlia/src/go_rng.rs]: It will only be used inside parlia.

File: crates/consensus/parlia/src/lib.rs (hunk 10 above)
Comment (pythonberg1997) [reply crates/consensus/parlia/src/lib.rs]: Used in `BscEVMProcesser.finalize`

File: crates/consensus/parlia/src/lib.rs (hunk 11 above)
Comment (pythonberg1997) [reply crates/consensus/parlia/src/lib.rs]: There're too many method of `parlia`, so I try to split them into different topics. Assemble system tx, abi encode/decode, check fork and others

File: crates/consensus/parlia/src/lib.rs (hunk 12 above)
Comment (pythonberg1997) [reply crates/consensus/parlia/src/lib.rs]: Sure

File: crates/revm/src/factory.rs (hunk 13 above)
Comment (pythonberg1997) [reply crates/revm/src/factory.rs]: Removed

File: crates/revm/src/processor.rs (hunk 16 above)
Comment (pythonberg1997) [reply crates/revm/src/processor.rs]: In fact there should never be empty block on BSC. Anyway, I remove this line.

File: crates/revm/src/processor.rs (hunk 17 above)
Comment (pythonberg1997) [reply crates/revm/src/processor.rs]: Fixd

**Pull request merged**